### DISCUSSION TITLE
Plugin system proposal

### DISCUSSION THREAD

**Discussion by advplyr** [advplyr/audiobookshelf#3744] [Ideas]
The following is a proposal to support plugins in Abs. [Corresponding draft PR](https://github.com/advplyr/audiobookshelf/pull/3745)

# Why
Users often request features that will likely not be a part of the core project for various reasons. This plugin system would allow extending Abs functionality to support those features.

See [these tagged issues](https://github.com/advplyr/audiobookshelf/issues?q=is%3Aopen+is%3Aissue+label%3A%22possible+plugin%22) for examples of feature requests that might be possible to build with this plugin system.

# How
Plugins are folders that are placed in `/metadata/plugins`.
The folders require 2 files (but may contain others):

1. index.js - Entry script
2. manifest.json - Includes all the details about the plugin like name, description, etc. as well as config & extensions (explained below)

Plugins are loaded and validated on server startup.

Each plugin will have a record in the `plugins` table. The primary purpose of this table is to allow plugins to store any config data it may need. The config data is stored in a `JSON` column.

# index.js

[Example minimal index.js file](https://github.com/advplyr/audiobookshelf/blob/plugin-implementation-demo/test/server/managers/plugins/Template/index.js)

Requires an `init` function that is called on server startup.
Supports optional `onAction` and `onConfigSave` functions.

Each function is passed a `context` object that contains (so far):
https://github.com/advplyr/audiobookshelf/blob/c8682c8456b838a4e0de6994225cc19484d87859/server/managers/PluginManager.js#L13-L21

`pluginInstance` is the Sequelize `Plugin` model instance for that plugin. This allows for easily reading and updating its config.

In this example `init` function the plugins config is updated using the `pluginInstance` and a Sequelize hook is setup to listen for updates to the `mediaProgresses` table.
https://github.com/advplyr/audiobookshelf/blob/c8682c8456b838a4e0de6994225cc19484d87859/test/server/managers/plugins/Example/index.js#L6-L23

# manifest.json

[Example manifest.json](https://github.com/advplyr/audiobookshelf/blob/plugin-implementation-demo/test/server/managers/plugins/Example/manifest.json)

The structure for this file is still a work in progress but the idea is that this file fully describes the plugin.

## UI Extensions
The manifest `extensions` array is a way to extend the Abs web client UI.
The purpose is to allow for triggering actions on specific pages and/or for specific entities.

Some key properties in the extension definition:

- name - used to identify the extension in the `onAction` function
- target - defines what page and where the extension is going
- label - shown to the user in the gui
- prompt - optionally defined to show a modal to collect additional information or to just confirm the action

The only extension allowed so far requires the `target` property of the extension to be `item.detail.actions`. This means the extension is added to the context menu on the item details page.
![image](https://github.com/user-attachments/assets/27729338-f8a9-4cf5-a3d6-083f8d974e96)

When the extension is pressed it will either a) load a prompt modal or b) make an API request that calls the plugins `onAction` function.

[This example extension](https://github.com/advplyr/abs-report-for-review-plugin/blob/master/Report%20for%20Review/manifest.json#L11-L57) has `prompt` defined with `formFields` to collect information.

![image](https://github.com/user-attachments/assets/6ead02cb-28bc-488d-a433-2c81fee2aa43)

The plugin receives the form input values in a `data` object. [Example onAction](https://github.com/advplyr/abs-report-for-review-plugin/blob/master/Report%20for%20Review/index.js#L20-L48)

Extensions can be restricted to only show for users with a certain account type using the `restrictToAccountTypes` array property.

## Config
The manifest `config` object is used to populate the plugins config page that will be available in the web client.

[In this example config](https://github.com/advplyr/abs-report-for-review-plugin/blob/master/Report%20for%20Review/manifest.json#L70-L87) 2 form fields are specified, a text input and a checkbox.

![image](https://github.com/user-attachments/assets/849e322e-b5b8-429c-8036-76f23f472f9d)
_Note: UI is currently using default HTML elements and will be updated to use Abs themed inputs._

When the config page is saved it makes an API request that will call the plugins `onConfigSave` function. [Example](https://github.com/advplyr/abs-report-for-review-plugin/blob/master/Report%20for%20Review/index.js#L50-L82)

The config page also links to the `documentationUrl` and `repositoryUrl` defined in the `manifest.json`. 

This page might be further expanded on to show the plugin version, changelog, etc.

# Localization

This is undecided and not fully utilized in the demo. The idea is to support a `localization` object in the manifest so that plugins can supply translations for all of their text.

# Enabling plugins

Plugins will be opt-in (via environment variable) until/if the plugin system is stable.
Set env variable `ALLOW_PLUGINS` to `1` or if running locally the `dev.js` file can export `AllowPlugins: true`

Once enabled, plugins are loaded on server startup and the config page will include the `Plugins` nav item.

# Test with the Report for Review plugin

To test the plugin system checkout the branch in the PR: https://github.com/advplyr/audiobookshelf/pull/3745

Download the latest zip file for the `Report for Review` plugin [here](https://github.com/advplyr/abs-report-for-review-plugin/releases) and extract the contents in `/metadata/plugins`. (this example plugin was inspired by the feature request https://github.com/advplyr/audiobookshelf/issues/3572)

_Note: The name of the folder for the plugin must match the name in the `manifest.json` file. This will probably change to prevent collisions but the idea is that the `/metadata/plugins` folder clearly shows which plugins are downloaded and they aren't obfuscated._

After starting the server with the `Report for Review` plugin you will see 2 new menu items on the item details page if you are an admin and 1 new menu item if you are a user.

You will also find the extension config page available on the plugins config page.

![image](https://github.com/user-attachments/assets/f1f5da20-3bcd-4cb7-9e3b-80d29fa6e1ff)

**Comment by TomK32** [comment]
I have an idea for a plugin* but before I sink any time into developing that plugin, is it still planned to add the plugin feature?

* one that lists all the available episodes of a podcast even if not downloaded yet. [You made clear that this feature wouldn't be for the app itself](https://github.com/advplyr/audiobookshelf/issues/2216#issuecomment-1763423433), but a plugin would solve the issue.

**Comment by moonblade** [comment]
Any updates on this?

**Reply by moonblade** [reply]
was planning a plugin, but not sure if this is abandoned or not.
